### DISCUSSION TITLE
clusters

### DISCUSSION THREAD

**Discussion by msiemionCalistapw** [moj-analytical-services/splink#706] [Q&A] [answered]
Is there a way to keep match probability as a column in clusters?? 

Thanks!

**Comment by ThomasHepworth** [comment, marked as answer]
Hi Michayla,

This isn't something we currently offer at the moment. The best you can do it tweak the threshold before performing the clustering.

This is a fairly simple request though and it's something I think we should be offering, so I'll try to do something on this today. I'll link the PR to here if/when it happens.

**Reply by msiemionCalistapw** [reply]
Hi Tom,

That would be fantastic, thank you so much!

**Reply by msiemionCalistapw** [reply]
Hello,

I was curious if you had made any headway here? I think I see a pull request that is possibly a match probability for clusters solution but not sure? 

Thanks,

Michayla

**Reply by ThomasHepworth** [reply]
Heya, I did a quick PR that joins together the pairwise comparisons and the clusters [which you can see here](https://github.com/moj-analytical-services/splink/pull/707).

I imagine this is what you were after. The pairwise comparisons (the output from `linker.predict()`) are what we use to create the clusters, so this allows you to gain some further insights into how everything has been slot together.

Robin has been off for the past week, so we haven't formally discussed the PR and whether we want to implement it yet.

<hr>

For the time being, if you want to recreate the view from my PR, you can use the following if you're in **duckdb** (with minor variations if you're using spark or athena) :
```python
from splink.duckdb.duckdb_linker import DuckDBLinker
import pandas as pd

# This df and settings obj can be found in the main splink repo if you
# want to test this demo code. 
from tests.basic_settings import get_settings_dict
df = pd.read_csv("./tests/datasets/fake_1000_from_splink_demos.csv")

settings = get_settings_dict()
settings["link_type"] = "link_only"

linker = DuckDBLinker(
    [df, df],
    settings,
)

df_predict = linker.predict()

cluster = 0.9
df_clusters = linker.cluster_pairwise_predictions_at_threshold(df_predict, cluster)

#### See from here for joining the two tables
from splink.unique_id_concat import (
    _composite_unique_id_from_edges_sql,
)
predict = df_predict.physical_name
representatives = df_clusters.physical_name

uid_cols = linker._settings_obj._unique_id_input_columns
uid_concat_repr_l = _composite_unique_id_from_edges_sql(uid_cols, None, "repr_l")
uid_concat_repr_r = _composite_unique_id_from_edges_sql(uid_cols, None, "repr_r")
uid_concat_l = _composite_unique_id_from_edges_sql(uid_cols, "l", "n")
uid_concat_r = _composite_unique_id_from_edges_sql(uid_cols, "r", "n")

filter_cond = "where cluster_id_l = cluster_id_r" # use if you want to view cluster links
# filter_cond = ""  # Use this line if you want all outputs from 

sql = f"""
    select
        n.*,
        repr_l.cluster_id as cluster_id_l,
        repr_r.cluster_id as cluster_id_r
    from {predict} as n
    left join
    {representatives} as repr_l
        on {uid_concat_l} = {uid_concat_repr_l}
    left join
    {representatives} as repr_r
        on {uid_concat_r} = {uid_concat_repr_r}
    {filter_cond}
    order by
        cluster_id_l, cluster_id_r
"""

# Execute our join
linker._con.execute(sql).fetch_df()
```


### Quick note on the above
If your original request was to enable the viewing of probabilities for _**all**_ ID comparisons in a cluster, just be aware that it's a little bit more involved and would require quite a bit of extra computation from splink.

As an example, if we were deduping and had IDs 1, 2, 3, 4 and 5 in a cluster, we'd need to rerun splink to check:
**ID_l** | **ID_r**
-- | --
1 | 2
1 | 3
1 | 4
1 | 5
2 | 3
2 | 4
2 | 5
3 | 4
3 | 5
4 | 5

i.e. you generate the cartesian product for each cluster.

The main problem with this, is that it can quickly start to create extremely large datasets. I don't believe there's much benefit in offering this and we'd prefer to avoid doing it and just let users explore this on their own if they are interested enough.

**Reply by msiemionCalistapw** [reply]
Hello,

This is perfect! I think I will stick with the match probability of pairwise predictions, as I do not have need to see all ID comparisons per cluster. Thank you so much.

Michayla

**Reply by msiemionCalistapw** [reply]
Is the math behind the overall match probability of the cluster an average or is it more involved?
